From f84addc63df91c0afcbb4bcb4464345f469b9bb3 Mon Sep 17 00:00:00 2001
From: Gayathri <gathreya@email.arizona.edu>
Date: Tue, 26 May 2015 10:41:02 -0700
Subject: [PATCH] RESKC-453: Do not show cancelled award docs in T&M history As
 an award administrator, I may need to cancel an award edit transaction
 instead of saving or submitting. The information may be incomplete,
 incorrect, or I may have selected the wrong award/node by mistake. I should
 be able to use the cancel button on the Actions panel to completely delete
 any updates or changes.

Steps:

Open existing award
Go to time and money and check history, note current version number and update timestamp/user
Return to award and click edit
Go to award actions and click cancel
Reopen award
Go to time and money and check history: cancelled award version will now be listed

Expected Result: all evidence of the cancelled document are removed from the Award, T&M, or Award Budget documents.
Currently, any cancelled documents are appearing as a Version in the T&M History panel.
---
 .../impl/TimeAndMoneyHistoryServiceImpl.java       | 34 +++++++++++++---------
 1 file changed, 20 insertions(+), 14 deletions(-)

diff --git a/coeus-impl/src/main/java/org/kuali/kra/timeandmoney/service/impl/TimeAndMoneyHistoryServiceImpl.java b/coeus-impl/src/main/java/org/kuali/kra/timeandmoney/service/impl/TimeAndMoneyHistoryServiceImpl.java
index 6dc5d4e..4f39464 100644
--- a/coeus-impl/src/main/java/org/kuali/kra/timeandmoney/service/impl/TimeAndMoneyHistoryServiceImpl.java
+++ b/coeus-impl/src/main/java/org/kuali/kra/timeandmoney/service/impl/TimeAndMoneyHistoryServiceImpl.java
@@ -22,8 +22,10 @@
 import org.apache.commons.collections4.CollectionUtils;
 import org.apache.commons.collections4.ListUtils;
 import org.apache.commons.lang3.StringUtils;
+import org.kuali.coeus.common.framework.version.VersionStatus;
 import org.kuali.coeus.sys.framework.controller.DocHandlerService;
 import org.kuali.coeus.sys.framework.gv.GlobalVariableService;
+import org.kuali.coeus.sys.framework.model.KcTransactionalDocumentBase;
 import org.kuali.kra.award.home.Award;
 import org.kuali.kra.award.home.AwardAmountInfo;
 import org.kuali.kra.award.version.service.AwardVersionService;
@@ -90,12 +92,13 @@
 	List<AwardVersionHistory> buildAwardVersionHistoryList(List<Award> awardVersionList, List<TimeAndMoneyDocument> docs) throws WorkflowException {
 		List<AwardVersionHistory> awardVersionHistoryCollection = new ArrayList<>();
 		for (Award award : awardVersionList) {
-			AwardVersionHistory awardVersionHistory = new AwardVersionHistory(award);
-			awardVersionHistory.setDocumentUrl(buildForwardUrl(award.getAwardDocument().getDocumentNumber()));
-			awardVersionHistory.setAwardDescriptionLine(award.getAwardDescriptionLine());
-			awardVersionHistory.setTimeAndMoneyDocumentHistoryList(getDocHistoryAndValidInfosAssociatedWithAwardVersion(docs, award.getAwardAmountInfos(), award));
-
-			awardVersionHistoryCollection.add(awardVersionHistory);
+            if (!award.getAwardSequenceStatus().equalsIgnoreCase(VersionStatus.CANCELED.toString())) {
+                AwardVersionHistory awardVersionHistory = new AwardVersionHistory(award);
+                awardVersionHistory.setDocumentUrl(buildForwardUrl(award.getAwardDocument().getDocumentNumber()));
+                awardVersionHistory.setAwardDescriptionLine(award.getAwardDescriptionLine());
+                awardVersionHistory.setTimeAndMoneyDocumentHistoryList(getDocHistoryAndValidInfosAssociatedWithAwardVersion(docs, award.getAwardAmountInfos(), award));
+                awardVersionHistoryCollection.add(awardVersionHistory);
+            }
 		}
 		return awardVersionHistoryCollection;
 	}
@@ -287,16 +290,18 @@ protected Boolean isInValidInfosCollection(TimeAndMoneyDocument doc, List<AwardA
 	protected void removeCanceledDocs(List<TimeAndMoneyDocument> docs) {
 		List<TimeAndMoneyDocument> tempCanceledDocs = new ArrayList<>();
 		for (TimeAndMoneyDocument doc : docs) {
-			if (doc.getDocumentHeader().hasWorkflowDocument()) {
-				if (doc.getDocumentHeader().getWorkflowDocument().isCanceled()) {
-					tempCanceledDocs.add(doc);
-				}
-			}
-		}
+            if (isDocCancelled(doc)) {
+                tempCanceledDocs.add(doc);
+            }
+        }
 		docs.removeAll(tempCanceledDocs);
 	}
-	
-	public List<TimeAndMoneyDocument> buildTimeAndMoneyListForAwardDisplay(Award award) throws WorkflowException {
+
+    protected boolean isDocCancelled(KcTransactionalDocumentBase doc) {
+        return doc.getDocumentHeader().hasWorkflowDocument() && doc.getDocumentHeader().getWorkflowDocument().isCanceled();
+    }
+
+    public List<TimeAndMoneyDocument> buildTimeAndMoneyListForAwardDisplay(Award award) throws WorkflowException {
 		Map<String, Object> fieldValues1 = new HashMap<>();
 		// get the award number.
 		fieldValues1.put(ROOT_AWARD_NUMBER, award.getAwardNumber());
@@ -309,6 +314,7 @@ protected void removeCanceledDocs(List<TimeAndMoneyDocument> docs) {
 	protected Map<String, String> getHashMapToFindActiveAward(String goToAwardNumber) {
 		Map<String, String> map = new HashMap<>();
 		map.put(AWARD_NUMBER, goToAwardNumber);
+
 		return map;
 	}
 
